core/storage/ipfs/graphsync/impl/network/message_queue.cpp

```diff
@@ -0,0 +1,115 @@
+/**
+ * Copyright Soramitsu Co., Ltd. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include "message_queue.hpp"
+
+#include <libp2p/connection/stream.hpp>
+
+#include <cassert>
+
+namespace fc::storage::ipfs::graphsync {
+
+  MessageQueue::MessageQueue(StreamPtr stream, FeedbackFn feedback)
+      : feedback_(std::move(feedback)) {
+    assert(feedback_);
+    assert(stream);
+    assert(!stream->isClosedForWrite());
+
+    state_.stream = std::move(stream);
+    state_.active = true;
+  }
+
+  const MessageQueue::State &MessageQueue::getState() const {
+    return state_;
+  }
+
+  void MessageQueue::enqueue(SharedData data) {
+    if (!data || data->empty() || !state_.active) {
+      // ignore if inactive and do nothing with empty messages for consistency
+      return;
+    }
+
+    if (state_.writing_bytes > 0) {
+      // waiting for async write completion, enqueue
+      state_.pending_bytes += data->size();
+      pending_buffers_.emplace_back(std::move(data));
+    } else {
+      // begin async write
+      beginWrite(std::move(data));
+    }
+  }
+
+  void MessageQueue::clear() {
+    state_.pending_bytes = 0;
+    pending_buffers_.clear();
+  }
+
+  void MessageQueue::close() {
+    clear();
+    state_.active = false;
+    state_.writing_bytes = 0;
+    state_.stream.reset();
+  }
+
+  void MessageQueue::dequeue() {
+    assert(state_.active);
+
+    if (!pending_buffers_.empty()) {
+      auto buffer = std::move(pending_buffers_.front());
+      pending_buffers_.pop_front();
+      state_.pending_bytes -= buffer->size();
+      beginWrite(std::move(buffer));
+    }
+  }
+
+  void MessageQueue::beginWrite(SharedData buffer) {
```

Comment:
Pass shared pointers by reference in cases, where we don't need to share ownership, is better
https://stackoverflow.com/questions/3310737/should-we-pass-a-shared-ptr-by-reference-or-by-value